File: Aqua Guard/frontend/app/src/main/java/com/aquaguard/presentation/auth/AuthViewModel.java
package com.aquaguard.presentation.auth;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.aquaguard.data.local.PreferencesManager;
import com.aquaguard.domain.model.User;
import com.aquaguard.domain.repository.AuthRepository;
import com.aquaguard.domain.repository.RepositoryCallback;

import javax.inject.Inject;
import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class AuthViewModel extends ViewModel {
    private final AuthRepository authRepository;
    private final PreferencesManager preferencesManager;
    
    private final MutableLiveData<AuthState> authStateLiveData = new MutableLiveData<>(AuthState.idle());

    @Inject
    public AuthViewModel(AuthRepository authRepository, PreferencesManager preferencesManager) {
        this.authRepository = authRepository;
        this.preferencesManager = preferencesManager;
    }

    public LiveData<User> getCurrentUser() {
        return authRepository.getCurrentUser();
    }

    public LiveData<AuthState> getAuthState() {
        return authStateLiveData;
    }

    public boolean isRememberMeEnabled() {
        return preferencesManager.isRememberMeEnabled();
    }

    public void setRememberMeEnabled(boolean value) {
        preferencesManager.setRememberMeEnabled(value);
    }

    public String getRememberedEmail() {
        return preferencesManager.getRememberedEmail();
    }

    public void signIn(String email, String password) {
        authStateLiveData.setValue(AuthState.loading());
        // Supports automatic registration of default admin credentials if they do not exist.
        authRepository.signInWithEmail(email, password, new RepositoryCallback<User>() {
            @Override
            public void onSuccess(User user) {
                if (preferencesManager.isRememberMeEnabled()) {
                    preferencesManager.setRememberedEmail(email);
                } else {
                    preferencesManager.setRememberedEmail("");
                }
                authStateLiveData.postValue(AuthState.success(user));
            }

            @Override
            public void onFailure(Exception e) {
                authStateLiveData.postValue(AuthState.error(e.getMessage() != null ? e.getMessage() : "Login failed"));
            }
        });
    }

    public void signUp(String email, String password, String displayName) {
        authStateLiveData.setValue(AuthState.loading());
        authRepository.signUpWithEmail(email, password, displayName, new RepositoryCallback<User>() {
            @Override
            public void onSuccess(User user) {
                authStateLiveData.postValue(AuthState.success(user));
            }

            @Override
            public void onFailure(Exception e) {
                authStateLiveData.postValue(AuthState.error(e.getMessage() != null ? e.getMessage() : "Registration failed"));
            }
        });
    }

    public void sendPasswordResetEmail(String email) {
        authStateLiveData.setValue(AuthState.loading());
        authRepository.sendPasswordResetEmail(email, new RepositoryCallback<Void>() {
            @Override
            public void onSuccess(Void result) {
                authStateLiveData.postValue(AuthState.idle());
            }

            @Override
            public void onFailure(Exception e) {
                authStateLiveData.postValue(AuthState.error(e.getMessage() != null ? e.getMessage() : "Failed to send reset email"));
            }
        });
    }

    public void signOut() {
        authRepository.signOut(new RepositoryCallback<Void>() {
            @Override
            public void onSuccess(Void result) {
                authStateLiveData.postValue(AuthState.idle());
            }

            @Override
            public void onFailure(Exception e) {
                // Ignore signOut failures
                authStateLiveData.postValue(AuthState.idle());
            }
        });
    }

    // Java Sealed-class equivalent
    public static class AuthState {
        public enum Status { IDLE, LOADING, SUCCESS, ERROR }
        private final Status status;
        private final User user;
        private final String message;

        private AuthState(Status status, User user, String message) {
            this.status = status;
            this.user = user;
            this.message = message;
        }

        public static AuthState idle() { return new AuthState(Status.IDLE, null, ""); }
        public static AuthState loading() { return new AuthState(Status.LOADING, null, ""); }
        public static AuthState success(User user) { return new AuthState(Status.SUCCESS, user, ""); }
        public static AuthState error(String msg) { return new AuthState(Status.ERROR, null, msg); }

        public Status getStatus() { return status; }
        public User getUser() { return user; }
        public String getMessage() { return message; }
    }
}

File: Aqua Guard/frontend/app/src/main/java/com/aquaguard/presentation/auth/AuthViewModel.kt
package com.aquaguard.presentation.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.aquaguard.data.local.PreferencesManager
import com.aquaguard.domain.model.User
import com.aquaguard.domain.repository.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class AuthViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val preferencesManager: PreferencesManager
) : ViewModel() {

    val currentUser: StateFlow<User?> = authRepository.currentUser
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), null)

    private val _authState = MutableStateFlow<AuthState>(AuthState.Idle)
    val authState: StateFlow<AuthState> = _authState.asStateFlow()

    var isRememberMeEnabled: Boolean
        get() = preferencesManager.isRememberMeEnabled
        set(value) {
            preferencesManager.isRememberMeEnabled = value
        }

    val rememberedEmail: String
        get() = preferencesManager.rememberedEmail

    fun signIn(email: String, password: String) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            authRepository.signInWithEmail(email, password)
                .onSuccess {
                    if (preferencesManager.isRememberMeEnabled) {
                        preferencesManager.rememberedEmail = email
                    } else {
                        preferencesManager.rememberedEmail = ""
                    }
                    _authState.value = AuthState.Success(it)
                }
                .onFailure {
                    _authState.value = AuthState.Error(it.message ?: "Login failed")
                }
        }
    }

    fun signUp(email: String, password: String, displayName: String) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            authRepository.signUpWithEmail(email, password, displayName)
                .onSuccess {
                    _authState.value = AuthState.Success(it)
                }
                .onFailure {
                    _authState.value = AuthState.Error(it.message ?: "Sign up failed")
                }
        }
    }

    fun sendPasswordResetEmail(email: String) {
        viewModelScope.launch {
            _authState.value = AuthState.Loading
            authRepository.sendPasswordResetEmail(email)
                .onSuccess {
                    _authState.value = AuthState.Idle
                }
                .onFailure {
                    _authState.value = AuthState.Error(it.message ?: "Failed to send reset email")
                }
        }
    }

    fun signOut() {
        viewModelScope.launch {
            authRepository.signOut()
            _authState.value = AuthState.Idle
        }
    }
}

sealed interface AuthState {
    object Idle : AuthState
    object Loading : AuthState
    data class Success(val user: User) : AuthState
    data class Error(val message: String) : AuthState
}
